components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -3692,5 +3249,628 @@ subroutine get_time_space_phys_variables( &
+   
+   logical :: log_qxpresent, qc_present, two_moment
+   integer :: k 
+   integer :: kbot, ktop, kdir, k_qxtop, k_qxbot, k_temp 
+   integer :: tmpint1 
+
+   real(rtype) :: dt_left 
+   real(rtype) :: prt_accum 
+   real(rtype) :: Co_max 
+   real(rtype) :: dt_sub
+   real(rtype), dimension(kts:kte) :: V_qc
+   real(rtype), dimension(kts:kte) :: V_nc 
+   real(rtype), dimension(kts:kte) :: flux_qx
+   real(rtype), dimension(kts:kte) :: flux_nx
+   real(rtype) :: fluxdiv_qx, fluxdiv_nx 
+ 
+   real(rtype) :: tmp1, tmp2, dum 
+
+   ktop = kts
+   kbot = kte
+   kdir = -1 
+   k_qxtop = kbot 
+   log_qxpresent = .false. 
+
+   !find top, determine qxpresent
+   do k = ktop,kbot,-kdir
+      if (qc(k).ge.qsmall) then
+         log_qxpresent = .true.
+         k_qxtop = k
+         exit
+      endif
+   enddo
+
+
+   qc_present: if (log_qxpresent) then
+
+
+      dt_left   = dt  !time remaining for sedi over full model (mp) time step
+      prt_accum = 0._rtype  !precip rate for individual category
+
+      !find bottom
+      do k = kbot,k_qxtop,kdir
+         if (qc(k).ge.qsmall) then
+            k_qxbot = k
+            exit
+         endif
+      enddo
+
+      two_moment: if (log_predictNc) then  !2-moment cloud:
+         substep_sedi_c2: do while (dt_left.gt.1.e-4_rtype)
+
+            Co_max = 0._rtype
+            V_qc = 0._rtype
+            V_nc = 0._rtype
+
+               kloop_sedi_c2: do k = k_qxtop,k_qxbot,-kdir
+
+               qc_notsmall_c2: if (qc_incld(k)>qsmall) then
+                  !-- compute Vq, Vn
+                  call get_cloud_dsd2(qc_incld(k),nc_incld(k),mu_c(k),rho(k),nu(k),dnu,   &
+                  lamc(k),lammin,lammax,tmp1,tmp2,lcldm(k))
+
+                  nc(k) = nc_incld(k)*lcldm(k)
+                  dum = 1._rtype/lamc(k)**bcn
+                  V_qc(k) = acn(k)*gamma(4._rtype+bcn+mu_c(k))*dum/(gamma(mu_c(k)+4._rtype))
+                  V_nc(k) = acn(k)*gamma(1._rtype+bcn+mu_c(k))*dum/(gamma(mu_c(k)+1._rtype))
+
+               endif qc_notsmall_c2
+               Co_max = max(Co_max, V_qc(k)*dt_left*inv_dzq(k))
+
+            enddo kloop_sedi_c2
+
+            !-- compute dt_sub
+            tmpint1 = int(Co_max+1._rtype)    !number of substeps remaining if dt_sub were constant
+            dt_sub  = min(dt_left, dt_left/float(tmpint1))
+
+            if (k_qxbot.eq.kbot) then
+               k_temp = k_qxbot
+            else
+               k_temp = k_qxbot-kdir
+            endif
+
+             !-- calculate fluxes
+            do k = k_temp,k_qxtop,kdir
+               flux_qx(k) = V_qc(k)*qc(k)*rho(k)
+               flux_nx(k) = V_nc(k)*nc(k)*rho(k)
+            enddo
+
+            !accumulated precip during time step
+            if (k_qxbot.eq.kbot) prt_accum = prt_accum + flux_qx(kbot)*dt_sub
+            !or, optimized: prt_accum = prt_accum - (k_qxbot.eq.kbot)*dt_sub
+
+            !-- for top level only (since flux is 0 above)
+            k = k_qxtop
```

Comment:
Coming up on this chunk of code for rain after looking at cloud.  It seems to me that the actual sedimentation calculations are always the same, given the inputs of V_qX, qX, V_nX and nX.  Could we make a general sedimentation routine that each species calls?

Reply:
Yes, we can and should as this is how it will be done in the C++ code. I might as well go ahead and add it as part of this PR.